utilities/src/serialize/impls.rs

```diff
@@ -398,6 +398,44 @@ impl<T: CanonicalDeserialize> CanonicalDeserialize for Vec<T> {
     }
 }
 
+impl<T: CanonicalDeserialize> CanonicalDeserialize for [T; 32] {
+    #[inline]
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let values = [(); 32]
+            .iter()
+            .map(|_| T::deserialize_with_mode(&mut reader, compress, Validate::No))
```

Comment:
If we use `array::map` instead of `Iterator::map`, then we can avoid allocation.

So something like this:

```rust
let values = [(); 32]
	.map(|_| T::deserialize_with_mode(&mut reader, compress, Validate::No));

for value in &values {
    // check that each value is error free
    let _ = value.as_ref()?;
}

let values = values.map(|r| r.unwrap());

if let Validate::Yes = validate {
    T::batch_check(values.iter());
}
Ok(values)
```